src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -241,18 +462,47 @@ private static MethodHandle bindCaller(MethodHandle target, Class<?> hostClass)
      * @return true if the reference is not null and points to an object produced by {@code asInterfaceInstance}
      */
     public static boolean isWrapperInstance(Object x) {
-        return x instanceof WrapperInstance;
+        return isWrapperClass(x.getClass());
     }
 
-    private static WrapperInstance asWrapperInstance(Object x) {
+    static boolean isWrapperClass(Class<?> cls) {
         try {
-            if (x != null)
-                return (WrapperInstance) x;
-        } catch (ClassCastException ex) {
+            return WRAPPER_TYPES.get(cls) != null;
+        } catch (IllegalArgumentException ex) {
+            return false;
         }
-        throw newIllegalArgumentException("not a wrapper instance");
     }
 
+    private static final ClassValue<Class<?>> WRAPPER_TYPES = new ClassValue<>() {
+        @Override
+        protected Class<?> computeValue(Class<?> type) {
+            MethodHandle originalTypeField;
+            try {
+                originalTypeField = new Lookup(type).findStaticGetter(type, ORIGINAL_TYPE_NAME, Class.class);
```

Comment:
If some random class `X` has the field "originalType" with value `I`, `isWrapperClass` would cause spinning a MH proxy class `I` as it calls `getProxyClassInfo` and return true.   `getProxyClassInfo` should be only called from `asInterfaceInstance`.

I think this can be done by having a `WeakHashMap<Class<?>, Boolean>` and `getProxyClassInfo` puts an entry with `TRUE` when a new proxy class is generated.